File: src/lowlevel/client/icefsoperators/icefsMknod.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:53
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsMknod.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsMknod(fuse_req_t fuseReq, fuse_ino_t parentInode,
                               const char *name, mode_t mode, dev_t rdev) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsMknodReq req;
      icefsgrpc::IcefsMknodRes res;
      grpc::ClientContext ctx;
      struct fuse_entry_param entry;

      req.set_parent_inode(parentInode);
      req.set_name(name);
      req.set_mode(mode);
      req.set_rdev(rdev);

      grpc::Status status = gRpcClient->DoIcefsMknod(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        IcefsGRpcFillFuseEntryParamIn(&entry, res.entry());
        fuse_reply_entry(fuseReq, &entry);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsMknodReq req;
      icefsthrift::IcefsMknodRes res;
      struct fuse_entry_param entry;

      req.__set_parent_inode(parentInode);
      req.__set_name(name);
      req.__set_mode(mode);
      req.__set_rdev(rdev);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsMknod(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        IcefsThriftFillFuseEntryParamIn(&entry, res.entry);
        fuse_reply_entry(fuseReq, &entry);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsMknod.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 07:50:13
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsMknod.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsMknod(parentFakeInode uint64, name string, mode uint32, rdev uint64, fuseEntryParamBuilder FuseEntryParamBuilder) (status int32, entry any) {
	var err error
	s.inodeCacheLock.RLock()
	parentInode := s.getIcefsInode(parentFakeInode)
	if parentInode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	parentInode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	err = syscall.Mknodat(parentInode.fd, name, mode, int(rdev))
	parentInode.inodeLock.Unlock()
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}
	entry, err = s.doIcefsLookUp(parentFakeInode, name, fuseEntryParamBuilder)
	status = icefserror.IcefsStdErrno(err)

	return
}

func (s *IcefsGRpcServer) DoIcefsMknod(ctx context.Context, req *pb.IcefsMknodReq) (*pb.IcefsMknodRes, error) {
	var res pb.IcefsMknodRes
	var entry any

	res.Status, entry = s.server.doIcefsMknod(req.ParentInode, req.Name, req.Mode, req.Rdev, GRpcFuseEntryParamBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Entry = entry.(*pb.FuseEntryParam)
	}

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsMknod(ctx context.Context, req *icefsthrift.IcefsMknodReq) (*icefsthrift.IcefsMknodRes, error) {
	var res icefsthrift.IcefsMknodRes
	var entry any

	res.Status, entry = s.server.doIcefsMknod(uint64(req.ParentInode), req.Name, uint32(req.Mode), uint64(req.Rdev), ThriftFuseEntryParamBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Entry = entry.(*icefsthrift.FuseEntryParam)
	}

	return &res, nil
}
